### PR TITLE
[wip] Jnagmp for modular multiply, gcd, and exactDivide. 

### PR DESCRIPTION
**Pull request by jacobwil**
Today I submitted [a pull request to Square's JNA-GMP](https://github.com/square/jna-gmp/pull/17) which adds modular multiplication, gcd, and exactDivide to JNA-GMP. This library lets us take advantage of the GNU Multiple Precision Arithmetic Library's high performance integer math functions. 

Pending acceptance of my PR into JNA-GMP (hopefully) and a release, I have modified the pom files to pull an updated JNA-GMP jar out of my personal github repo. I don't intend for the pom files to be accepted into master in this state but wanted to make the wip available for everyone to try if they want. 

Benchmark results:

    Benchmark                                   Mode  Cnt        Score        Error  Units
    ExactDivideBenchmark.testWithGMP           thrpt  200   110961.159 ±   1071.711  ops/s
    ExactDivideBenchmark.testWithoutGMP        thrpt  200    53641.770 ±    267.455  ops/s
    GCDBenchmark.testWithGMP                   thrpt  200    24121.188 ±    144.105  ops/s
    GCDBenchmark.testWithoutGMP                thrpt  200     1243.952 ±      7.829  ops/s
    ModBenchmark.testWithGMP                   thrpt  200    93148.587 ±    762.959  ops/s
    ModBenchmark.testWithoutGMP                thrpt  200  2300046.853 ± 267769.213  ops/s
    ModularInverseBenchmark.testWithGMP        thrpt  200     9891.903 ±     38.648  ops/s
    ModularInverseBenchmark.testWithoutGMP     thrpt  200      390.049 ±      1.966  ops/s
    ModularMultiplyBenchmark.testWithGMP       thrpt  200    36154.112 ±    219.346  ops/s
    ModularMultiplyBenchmark.testWithoutGMP    thrpt  200    15777.195 ±    100.081  ops/s
    MultiplyBenchmark.testWithGMP              thrpt  200   102739.040 ±    522.167  ops/s
    MultiplyBenchmark.testWithoutGMP           thrpt  200   229137.717 ±    658.034  ops/s
    PaillierBenchmark.testWithGMP              thrpt  200       26.943 ±      0.122  ops/s
    PaillierBenchmark.testWithGMPConstantTime  thrpt  200       19.695 ±      0.078  ops/s
    PaillierBenchmark.testWithoutGMP           thrpt  200        4.573 ±      0.012  ops/s

Since modularMultiply is a major operation in the wideskies algorithm the 1.3x speedup we see is a huge win.

### CODE REVIEW

File: pom-with-benchmarks.xml [hunk 1]
```diff
@@ -67,7 +67,7 @@
 			<id>mvn-public</id>
```
Comment (ellisonanne) [inline pom-with-benchmarks.xml]: I am assuming that most of the changes are due to a merge from an older pom file on your side...

File: pom.xml [hunk 2]
```diff
@@ -95,6 +95,18 @@
         <pirk.forkCount>1C</pirk.forkCount>
         <pirk.reuseForks>true</pirk.reuseForks>
     </properties>
```
Comment (ellisonanne) [inline pom.xml]: I'm fine with this for now in order to leverage the performance gain - however, we will need another long term solution if your PR to JNA-GMP is not rolled into a release in a timely manner.

File: src/main/java/org/apache/pirk/encryption/IntegerMathAbstraction.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  {
+    // Load the configuration
+    reloadConfiguration();
+  }
+
+  /**
+   * Reloads the configuration options for this class. They control which implementations are used for integer
+   * math: GMP or BigInteger.
+   */
+  public static void reloadConfiguration()
```
Comment (ellisonanne) [inline src/main/java/org/apache/pirk/encryption/IntegerMathAbstraction.java]: These can be replaced with SystemConfiguration.getBooleanProperty

**Comment by ellisonanne** [comment]
Looks good - I made a few minor comment inline with the code above - when do you think that it will be ready to go from a [WIP] PR to ready-to-merge?

**Comment by jacobwil** [comment]
Thanks for the feedback Ellison Anne. I've been on a merge-with-master conflict resolution treadmill for a bit. Before I suggest this PR is ready for final merging I'll grab the latest poms and surgically make the version changes. 

When the PR will be ready to merge depends on JNA-GMP's response to my PR. As things happen I'll post more comments here.

**Comment by jacobwil** [comment]
I am unable to complete this pull request because JNA-GMP has not done a release with the new features and decided that composed functions like modular multiply were out of their scope.

**Pull request closed without merging**